File: solutions/kotlin/allergies/1/src/main/kotlin/Allergies.kt
class Allergies(val score: Int) {
    private val allergenScores = mapOf(
        Allergen.EGGS to 1,
        Allergen.PEANUTS to 2,
        Allergen.SHELLFISH to 4,
        Allergen.STRAWBERRIES to 8,
        Allergen.TOMATOES to 16,
        Allergen.CHOCOLATE to 32,
        Allergen.POLLEN to 64,
        Allergen.CATS to 128
    )
    
    fun getList(): List<Allergen> = buildList {
        for ((allergen, value) in allergenScores) {
            if (score and value != 0) {
                add(allergen)
            }
        }
    }
    
    fun isAllergicTo(allergen: Allergen): Boolean =
        allergenScores[allergen]?.let { value ->
            score and value != 0
        } ?: false
}

File: solutions/zig/allergies/5/allergies.zig
const std = @import("std");
const EnumSet = std.EnumSet;

pub const Allergen = enum(u4) {
    eggs,
    peanuts,
    shellfish,
    strawberries,
    tomatoes,
    chocolate,
    pollen,
    cats,
};

pub fn isAllergicTo(score: usize, allergen: Allergen) bool {
    return score >> @intFromEnum(allergen) & 1 == 1;
}

pub fn initAllergenSet(score: usize) EnumSet(Allergen) {
    return .{ .bits = .{ .mask = @truncate(score) } };
}
